src/sdk/services/oauthService.ts

```diff
@@ -2,33 +2,48 @@ import { ManifestApi } from '../../manifest/ManifestApi';
 import runtime from '../RuntimeContext';
 
 class OAuthService {
-  public openPopup = () => {
+  public openPopup = (redirectUri?: string) => {
     if (!runtime.application.api) {
-      throw new Error(
-        "Can't open auth window for addon which has no api configuration in its manifest"
-      );
+      throw new Error("Can't open auth window for addon which has no api configuration in its manifest");
     }
 
-    const authorizeUrl = this.getOAuthAuthorizeUrl(runtime.application.api);
+    const authorizeUrl = this.getOAuthAuthorizeUrl(runtime.application.api, redirectUri);
     this.showPopup(authorizeUrl, 800, 600);
   };
 
   private getOAuthHost = () => {
     const originUrl = new URL(runtime.origin);
 
-    const regex = /https:\/\/(\w+?)\./.exec(originUrl.origin);
+    const regex = /https:\/\/([-\w]+?)\./.exec(originUrl.origin);
     if (!regex) {
       throw new Error('Invalid runtime origin url:' + originUrl.origin);
     }
 
     return originUrl.origin.replace(regex[1], 'accounts');
   };
 
-  private getOAuthAuthorizeUrl = (api: ManifestApi) => {
+  private getOAuthAuthorizeUrl = (api: ManifestApi, redirectUri?: string) => {
     const host = this.getOAuthHost();
     const scopes = encodeURIComponent(api.scopes.join(' '));
-    const redirectUri = encodeURIComponent(api.redirectUri);
-    return `${host}/oauth/authorize?client_id=${api.applicationId}&redirect_uri=${redirectUri}&response_type=code&scope=${scopes}`;
+    const selectedRedirectUri = encodeURIComponent(this.selectRedirectUri(api, redirectUri));
+    const clientId = encodeURIComponent(api.client.id || api.applicationId || '');
+    return `${host}/oauth/authorize?client_id=${clientId}&redirect_uri=${selectedRedirectUri}&response_type=code&scope=${scopes}`;
+  };
+
+  private selectRedirectUri = (api: ManifestApi, redirectUri?: string): string => {
+    if (redirectUri === undefined) {
+      if (api.redirectUris?.length) {
+        return api.redirectUris[0];
+      }
+
+      return api.redirectUri ?? '';
```

Comment:
If api.redirectUri is undefined here you will return empty string - will this be valid value? in L46 you throw and not return empty string?

Reply:
In L46 a parameter provided by a developer doesn't match any of redirectUris, that's why we throw. Here it's different.